### DISCUSSION TITLE
Declarative model as separate class with dataclass mapping

### DISCUSSION THREAD

**Discussion by sh-at-cs** [sqlalchemy/sqlalchemy#8483] [Usage Questions] [answered]
I read the [docs on dataclass integration](https://docs.sqlalchemy.org/en/14/orm/dataclasses.html) and have been wondering:

Is there a way to get the legibility of database models defined as classes (versus lists of `Column`s as in the imperative style) while simultaneously keeping the dataclass definitions themselves "pure", i.e. not having to insert any SQLAlchemy code into them?<sup>1</sup>

So it would look something like:

```python
# pure_models.py

@dataclass
class User:
  name: str
  email: str
```

```python
# sql_models.py

class User(Base):
  name = Column(...)
  email = Column(...)

mapper_registry.some_method_for_associating_them(pure_models.User, User)
```

So the declarative model here would act purely as syntactic sugar for what would normally be an imperative definition (`some_method_for_associating_them` would be very similar to `map_imperatively`).

I tried to see if something like this could be hacked together by just going through the attributes on a declarative model class and turning them into `Column`s for an imperative definition, but there is too much magic going on that e.g. transforms them into `InstrumentedAttribute`s by the time the class is created at runtime, so I can't access the arguments to `Column` anymore.

Any ideas, suggestions, 3rd party projects, ..?

-----

<sup>1</sup>I know that at runtime SQLAlchemy injects its `_sa_instance_state` and so on into mapped dataclass _instances_ anyway, but I'm only concerned with the class definitions as visible in the code here.

**Comment by zzzeek** [comment, marked as answer]
That link you provided seems to be the current 1.4 docs.   That stuff is already legacy.     We have a much nicer dataclass integration coming up which you can see at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping .  

now from that, to go to your "pure" model, you would build a function that takes the "pure" dataclass, scans it for `: str` whatever, change them to `Mapped[str]`, build up the new class and send it off to registry.mapped_as_dataclass().

from my POV, such an approach is not industrial scale, since it has no way to indicate fields that should be mapped vs. fields that are just plain fields, has no way to specify any kinds of options for the columns / relationships / ORM directives themselves, doesn't even have a way of indicating what column(s) form the primary key.  that would be all things you'd need to invent.

**Reply by sh-at-cs** [reply]
Thanks for the extremely fast response!

Also thank you for pointing me to the docs of the 2.0 dataclass integration, I'd read about it in some other tickets but wasn't sure what the status of those was.

Regarding your suggested approach, that seems to be about having the dataclass definition as the "single source of truth" for both the "pure" and the DB model, but what I was talking about was closer to how it already works with imperative mappings (which are not deprecated in 2.0 according to the docs you linked), i.e. you do define your DB model separately from the dataclass model, _except_ that this would be done in the form of classes with `x = Column(...)` fields instead of a list of columns (so really just syntactic sugar for the latter).

Now that I think about it, I guess this should actually be fairly easy to build from scratch without involving SQLAlchemy's declarative functionality at all... I'll see what I can do, thanks!

**Reply by zzzeek** [reply]
OK well, make a separate Base, then define old-school declarative models using `x = Column(...)`, then grab that ``__table__`` and map it to your dataclass.   That's fine, you can do that.   

(from my pov seems kind of nutty, but you do you!)

**Reply by sh-at-cs** [reply]
The separate base +  `__table__` approach is exactly what I was looking for, thanks a lot!!

EDIT:

And for anyone coming across this in the future: In fact, you don't even need `__table__`, as `map_imperatively` takes any `FromClause` as its 2nd argument and DB model classes fall into that category, so you can just pass the class directly.

So `some_method_for_associating_them` I was looking for *is* `map_imperatively` and I should have just tried it :facepalm:
